dbms/src/Storages/Page/V3/PageDirectory.h

```diff
@@ -82,48 +82,172 @@ class PageDirectorySnapshot : public DB::PageStorageSnapshot
 };
 using PageDirectorySnapshotPtr = std::shared_ptr<PageDirectorySnapshot>;
 
+// This is not a thread safe class
+// It must be used with outer synchronization
+class MultiVersionRefCount
+{
+public:
+    static MultiVersionRefCount * createFrom(const PageVersion & ver, Int64 ref_count)
+    {
+        if (ref_count == 1)
+        {
+            return nullptr;
+        }
+        return appendRefCount(nullptr, ver, ref_count);
+    }
+
+    static MultiVersionRefCount * copy(MultiVersionRefCount * container)
+    {
+        if (container == nullptr)
+        {
+            return nullptr;
+        }
+        auto * new_container = new MultiVersionRefCount();
+        RUNTIME_CHECK(new_container != nullptr);
+        for (const auto & [ver, ref_count] : container->ref_map)
+        {
+            new_container->ref_map.emplace(ver, ref_count);
+        }
+        return new_container;
+    }
+
+    static void destroy(MultiVersionRefCount * container)
+    {
+        delete container;
+    }
+
+    static void decrLatestRefCountInSnap(MultiVersionRefCount * container, UInt64 snap_seq, Int64 deref_count)
```

Comment:
I wonder do we really need to literate all the versions and decrease ref_count? How about make `ref_map` an update history instead of update result. I.e. We postpone the final ref_count of each version to the actual call of `getRefCountInSnap` and `getLatestRefCount`. So that `decrLatestRefCountInSnap` will be much simpler.

Reply:
And how about making `ref_map`  an `std::vector` instead of `std::map`. Because `vector` is faster and size efficient than `map`. And for the rare case that we need to do random update and truncate smaller versions, just create an new `vector` and do copy.

Reply:
To maintain an update history will just make the code a lot complicated:
Consider the following operations:
1. put 1 at version 10
2. ref 2 -> 1 at version 11
3. ref 3 -> 1 at version 12
And now the version history becomes
```
version -> ref count
10 -> 1
11 -> 2
12 -> 3
```
Operations continue:
1. del 3 at version 13
2. del 2 at version 14
Then gc happens:
1. deref 2
2. deref 3
After the first deref op happens, the version history becomes
```
10 -> 1
11 -> 2
12 -> 3
14 -> 2
```
Then after the second deref op, the version history should become:
```
10 -> 1
11 -> 2
12 -> 3
13 -> 2
14 -> 1
```
Note the ref count of **version 14 changes from 2 to 1** after the second deref which make it essentially not an update history.
The reason why this problem happens is that if we deref at gc time, the delete operations of ref id doesn't come at version order.

And then a natural idea is to deref at write time. But because when do full gc, we may rewrite a ref id to a normal page id, so we will still meet a similar situation like the above at gc time.

Reply:
Fixed.